### ISSUE TITLE
ErrImagePull should be a retryable error

### ISSUE THREAD

**Issue by johnSchnake** [vmware-tanzu/sonobuoy#969]
**What steps did you take and what happened:**
Ran a sonobuoy run with systemd-logs running on 6 nodes. One node temporarily entered the errimagepull state but recovered and ran. However, the results were not processed by the aggregator because it was deemed to be a final error.

**What did you expect to happen:**
ErrImagePull and other potentially transient errors should be retryable. Perhaps all results should be overwriteable? Currently the code ensures duplicates are tossed out, but I'm not sure if there isn't a reason we don't just keep the most recent. That would avoid this problem coming back when other types of retryable failures come up.

**Anything else you would like to add:**
The other solution (besides allowing blanket retries) is to allow the error result to specify somehow that the error is retryable. I think this can may get complicated if plugins themselves have retryable failure modes (and not just the k8s level errors where the job/daemonset enters something like ErrImagePull).

Open to other ideas.

As a side note, there may be a bug in the logic checking for duplicate results; only one node reported this state but subsequently, every node was getting 409s when uploading results (duplicate results) and they just kept retrying. I'm not even exactly sure how they got results in that case: 
 - if the first try succeeded in passing the results, why did the workers keep retrying and getting 409s?

The logs in question were a bit odd since the `run.log` on the aggregator shows redundant results multiple times for the nodes but each worker shows it at most once :
```
{"level":"info","msg":"Running plugin","plugin":"e2e","time":"2019-10-18T15:32:24Z"}
{"level":"info","msg":"Running plugin","plugin":"systemd-logs","time":"2019-10-18T15:32:24Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"minion-1","plugin_name":"systemd-logs","time":"2019-10-18T15:32:55Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"master-0","plugin_name":"systemd-logs","time":"2019-10-18T15:32:58Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"master-2","plugin_name":"systemd-logs","time":"2019-10-18T15:32:58Z"}
{"level":"info","msg":"received internal aggregator result","node":"master-1","plugin_name":"systemd-logs","time":"2019-10-18T15:33:07Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"minion-2","plugin_name":"systemd-logs","time":"2019-10-18T15:33:16Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"minion-0","plugin_name":"systemd-logs","time":"2019-10-18T15:33:29Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"master-1","plugin_name":"systemd-logs","time":"2019-10-18T15:33:56Z"}
{"level":"error","msg":"Result processing error (409): result systemd-logs/master-1 already received","time":"2019-10-18T15:33:56Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"minion-1","plugin_name":"systemd-logs","time":"2019-10-18T16:32:57Z"}
{"level":"error","msg":"Result processing error (409): result systemd-logs/minion-1 already received","time":"2019-10-18T16:32:57Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"master-0","plugin_name":"systemd-logs","time":"2019-10-18T16:33:01Z"}
{"level":"error","msg":"Result processing error (409): result systemd-logs/master-0 already received","time":"2019-10-18T16:33:01Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"master-2","plugin_name":"systemd-logs","time":"2019-10-18T16:33:02Z"}
{"level":"error","msg":"Result processing error (409): result systemd-logs/master-2 already received","time":"2019-10-18T16:33:02Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"minion-2","plugin_name":"systemd-logs","time":"2019-10-18T16:33:19Z"}
{"level":"error","msg":"Result processing error (409): result systemd-logs/minion-2 already received","time":"2019-10-18T16:33:19Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"minion-0","plugin_name":"systemd-logs","time":"2019-10-18T16:33:32Z"}
{"level":"error","msg":"Result processing error (409): result systemd-logs/minion-0 already received","time":"2019-10-18T16:33:32Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"master-1","plugin_name":"systemd-logs","time":"2019-10-18T16:33:59Z"}
{"level":"error","msg":"Result processing error (409): result systemd-logs/master-1 already received","time":"2019-10-18T16:33:59Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"minion-1","plugin_name":"systemd-logs","time":"2019-10-18T17:32:59Z"}
{"level":"error","msg":"Result processing error (409): result systemd-logs/minion-1 already received","time":"2019-10-18T17:32:59Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"master-0","plugin_name":"systemd-logs","time":"2019-10-18T17:33:03Z"}
{"level":"error","msg":"Result processing error (409): result systemd-logs/master-0 already received","time":"2019-10-18T17:33:03Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"master-2","plugin_name":"systemd-logs","time":"2019-10-18T17:33:04Z"}
{"level":"error","msg":"Result processing error (409): result systemd-logs/master-2 already received","time":"2019-10-18T17:33:04Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"minion-2","plugin_name":"systemd-logs","time":"2019-10-18T17:33:21Z"}
{"level":"error","msg":"Result processing error (409): result systemd-logs/minion-2 already received","time":"2019-10-18T17:33:21Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"minion-0","plugin_name":"systemd-logs","time":"2019-10-18T17:33:33Z"}
{"level":"error","msg":"Result processing error (409): result systemd-logs/minion-0 already received","time":"2019-10-18T17:33:33Z"}
{"client_cert":"systemd-logs","level":"info","msg":"received aggregator request","node":"master-1","plugin_name":"systemd-logs","time":"2019-10-18T17:34:01Z"}
{"level":"error","msg":"Result processing error (409): result systemd-logs/master-1 already received","time":"2019-10-18T17:34:01Z"}
{"client_cert":"e2e","level":"info","msg":"received aggregator request","plugin_name":"e2e","time":"2019-10-18T17:42:02Z"}
{"level":"info","msg":"Last update to annotations on exit","time":"2019-10-18T17:42:02Z"}
```

Where the worker for master-1 showed:
```
time="2019-10-18T17:33:03Z" level=info msg="Detected done file, transmitting result file" resultFile=/tmp/results/systemd_logs
time="2019-10-18T17:33:03Z" level=error msg="gathering results: got a 409 response when dialing aggregator to https://[10.44.0.4]:8080/api/v1/results/by-node/master-0/systemd-logs"
```

The logging should make this situation more clear; this timestamp corresponds to 2hr after the start of the run.

Perhaps the situation was that it took 2hr for the errimagepull to go away but in that case _every_ node should have been in the same boat: getting 409s, no results, a final status of failed.

Instead, it seemed like the first errImagePull was recorded and the others were allowed to continue. But _something_ had to be recorded for the other nodes due to the 409s they were getting.

So, in summary:
 - errors discovered via `pluign.Monitor()` are treated as final errors even though they may not be
 - a bug in the monitoring logic for daemonsets seems to short circuit the handling for some nodes when an error is encountered
 - unsure why we dont allow 'duplicate' results and just store the last one
 - logging should be improved when sending/getting results to help understand these situations

**Comment by johnSchnake** [comment]
From the realization that spurred #970, lots of the confusion here is removed. The root issue is still valid: one pod of the daemonset got a errimagepull error and, even though it recovered and tried to send results, was denied.

**Comment by johnSchnake** [comment]
Having had a think on this, I am a little less sure about allowing all results to be overwritten. That may still be the way to go, but the downside is that we lose finality: even after every plugin reports results (errors or otherwise) they _may_ still decide to post newer results.

That would make some awkward behavior because of a temporal dependency between plugins: 
 - if plugin A reports error
 - waits 5 min 
 - tries to submit success 
 - it will have different results if plugin B finishes before or after that 5m mark: if both plugins have results sonobuoy will be 'done' and move on, if B is running we'd be letting A submit new results

**Comment by johnSchnake** [comment]
Adding to the next milestone for consideration.
